### ISSUE TITLE
Windows WSL2 is super slow.

### ISSUE THREAD

**Issue by prodchia** [madMAx43v3r/chia-plotter#133]
I am running it on a 40 cores Xeon V4 server with 256GB ram and 3.3T NVMe. I haveallocated 124 GB and 20 cores to the ubuntu subprocess.I am not seeing any significant performance gains.

user@server:~/chia/chia-plotter/build$ ./chia_plot -r 16 -t /mnt/d/ -d /mnt/m/test/ -2 /mnt/ramdisk/ -p <privatekey> -f <farmerkey>
Final Directory: /mnt/m/test/
Number of Plots: 1
Number of Threads: 16
Number of Buckets: 2^7 (128)
Pool Public Key:   
Farmer Public Key: 
Working Directory:   /mnt/d/
Working Directory 2: /mnt/ramdisk/
Plot Name: plot-k32-2021-06-10-03-08-0874b22b115299c30c0f1e92d9bd633ac79f69ab0a1fd132cb638c3bf6d78d27
[P1] Table 1 took 17.4673 sec
[P1] Table 2 took 620.944 sec, found 4294969977 matches
[P1] Lost 2439 matches due to 32-bit overflow.
[P1] Table 3 took 680.568 sec, found 4294907592 matches
[P1] Table 4 took 741.8 sec, found 4294925654 matches
[P1] Table 5 took 727.111 sec, found 4294943269 matches
[P1] Table 6 took 738.086 sec, found 4294903133 matches
[P1] Table 7 took 657.984 sec, found 4294864107 matches
**Phase 1 took 4183.98 sec**
[P2] max_table_size = 4294969977
[P2] Table 7 scan took 11.9844 sec
[P2] Table 7 rewrite took 38.0404 sec, dropped 0 entries (0 %)
[P2] Table 6 scan took 78.9933 sec
[P2] Table 6 rewrite took 267.604 sec, dropped 581262153 entries (13.5338 %)
[P2] Table 5 scan took 81.285 sec
[P2] Table 5 rewrite took 256.919 sec, dropped 761973332 entries (17.7412 %)
[P2] Table 4 scan took 82.9823 sec
[P2] Table 4 rewrite took 251.883 sec, dropped 828809403 entries (19.2974 %)
[P2] Table 3 scan took 101.163 sec
[P2] Table 3 rewrite took 251.128 sec, dropped 855037061 entries (19.9082 %)
[P2] Table 2 scan took 103.053 sec
[P2] Table 2 rewrite took 247.11 sec, dropped 865527590 entries (20.1521 %)
**Phase 2 took 1803.79 sec**
Wrote plot header with 268 bytes
[P3-1] Table 2 took 127.803 sec, wrote 3429442387 right entries
[P3-2] Table 2 took 1242.73 sec, wrote 3429442387 left entries, 3429442387 final
[P3-1] Table 3 took 182.72 sec, wrote 3439870531 right entries
[P3-2] Table 3 took 1263.4 sec, wrote 3439870531 left entries, 3439870531 final
[P3-1] Table 4 took 178.663 sec, wrote 3466116251 right entries
[P3-2] Table 4 took 1344.06 sec, wrote 3466116251 left entries, 3466116251 final
[P3-1] Table 5 took 252.212 sec, wrote 3532969937 right entries
[P3-2] Table 5 took 1311.65 sec, wrote 3532969937 left entries, 3532969937 final
[P3-1] Table 6 took 287.077 sec, wrote 3713640980 right entries
[P3-2] Table 6 took 1398.44 sec, wrote 3713640980 left entries, 3713640980 final
[P3-1] Table 7 took 62.3986 sec, wrote 4294864107 right entries
[P3-2] Table 7 took 1619.36 sec, wrote 4294864107 left entries, 4294864107 final
**Phase 3 took 9276.92 sec,** wrote 21876904193 entries to final plot
[P4] Starting to write C1 and C3 tables

I suspect this would be because of WSL2 slowness in interacting with the NTFS file system. Any pointer will be appreciated.

**Comment by madMAx43v3r** [comment]
You can try the new docker image: https://github.com/madMAx43v3r/chia-plotter/pull/102

**Comment by prodchia** [comment]
That doesn't look promising either.

https://github.com/madMAx43v3r/chia-plotter/issues/74#issuecomment-858568466

**Comment by madMAx43v3r** [comment]
Is this latest master? I improved P3-2 quite a bit, it should now be about 2x faster than P3-1.

**Comment by madMAx43v3r** [comment]
But yeah, NTFS could be another bottleneck

**Comment by prodchia** [comment]
It's about a day old. Let me compile the latest master and try again. Thanks for the pointer.

**Comment by delucca** [comment]
Also, don't forget to check the Docker performance limits: https://docs.docker.com/config/containers/resource_constraints/

Your Docker installation could be limiting your resources.

**Comment by VertiHydro** [comment]
@madMAx43v3r it's not an NTFS problem on WSL2: WSL2 literally maps the host drives as network shares to the Linux guest, resulting in abysmal performance: https://github.com/microsoft/WSL/issues/4197 (10x worse than WSL1) so please don't use WSL2 for plotting

**Comment by delucca** [comment]
So, according to https://github.com/moby/moby/issues/31604), when Docker for Windows, it is actually using a Hyper-V isolation model (and process model is not accessible in Win 10 scenario). And that VM has a memory limit of 1gb

So, you need to set the RAM to your Docker container manually with the `-m` argument :)

**Comment by VertiHydro** [comment]
Explanation on WSL2 vs WSL1 vs Native filesystems here if anyone is interested:
https://github.com/microsoft/WSL/issues/4197#issuecomment-604592340

**Comment by MinerGreggy** [comment]
yeah the I/O on WSL2 will max at 38mb its not worth to use it in WSL2

**Comment by delucca** [comment]
Just confirmed that with Docker you can run without CPU and memory limits. You just need to provide the `--cpus` and `-m` flags. So, Docker is way faster than WSL in Windows.

**Comment by Smallthing** [comment]
I already reply in this [https://github.com/madMAx43v3r/chia-plotter/issues/41](url)
you can plotting in wsl2 with ext4 
and modify wsl2's limit .
you can mount a ext4 or xfs raw disk with wsl2  [https://github.com/madMAx43v3r/chia-plotter/issues/41](url)

**Comment by PresidentCake** [comment]
I think you can try wsl1, it's mush better than wsl2

**Comment by PeratX** [comment]
WSL1 is more friendly with these filesystem sensitive program like chia plotter.

**Comment by feder-cr** [comment]
How can i install wsl1 on windows server?
@PeratX @PresidentCake @Smallthing @MinerGreggy @VertiHydro @

**Comment by VertiHydro** [comment]
> How can i install wsl1 on windows server?
> @PeratX @PresidentCake @Smallthing @MinerGreggy @VertiHydro @

I think so - I haven't tried it, but [this](https://docs.microsoft.com/en-us/windows/wsl/install-on-server) support article seems to indicate that you can if you're running Server 2019 build 1709 or later.

**Comment by feder-cr** [comment]
@VertiHydro Yes but is wsl1 or wsl2?

**Comment by VertiHydro** [comment]
> @VertiHydro Yes but is wsl1 or wsl2?

I believe it is WSL1 because it doesn't say anything about WSL2. Anyway, you can just install it and then run the command `wsl -l -v` to find out what version you're running.

**Comment by feder-cr** [comment]
unfortunately this command is not recognized on Windows server! @VertiHydro

**Comment by VertiHydro** [comment]
> unfortunately this command is not recognized on Windows server! @VertiHydro

have you installed WSL yet?

**Comment by feder-cr** [comment]
Yes

**Comment by VertiHydro** [comment]
> > unfortunately this command is not recognized on Windows server! @VertiHydro
> 
> have you installed WSL yet?

Nevermind I see that command will not work on WSL1 - give me a minute

**Comment by VertiHydro** [comment]
> Yes

Ok got it - open your preferred WSL distribution (for me Ubuntu) and type `uname -a` and post the result here.

**Comment by feder-cr** [comment]
Linux WIN-F85CP9CKAN0 4.4.0-17763-Microsoft #379-Microsoft Wed Mar 06 19:16:00 PST 2019 x86_64 x86_64 x86_64 GNU/Linux

@VertiHydro

**Comment by VertiHydro** [comment]
> Linux WIN-F85CP9CKAN0 4.4.0-17763-Microsoft #379-Microsoft Wed Mar 06 19:16:00 PST 2019 x86_64 x86_64 x86_64 GNU/Linux
> 
> @VertiHydro

Yep OK that is WSL1 - WSL2 would show "microsoft-standard" instead of "Microsoft" and the kernel version would be 4.17.something not 4.4.0. So, good - you are already on WSL1.

**Comment by feder-cr** [comment]
Thanks bro, but only 1 cpu of 2 was used!!
@VertiHydro

**Comment by VertiHydro** [comment]
> Thanks bro, but only 1 cpu of 2 work!!
> @VertiHydro

Hmmm - that appears to be a very common problem at the moment - I don't believe anybody knows why it occurs. You can search around the issues for other people with the same problem - I am afraid I do not know of a fix at present.

**Comment by altendky** [comment]
> I already reply in this [https://github.com/madMAx43v3r/chia-plotter/issues/41](url)
> you can plotting in wsl2 with ext4
> and modify wsl2's limit .
> you can mount a ext4 or xfs raw disk with wsl2 [https://docs.microsoft.com/en-us/windows/wsl/wsl2-mount-disk](url)
> [https://devblogs.microsoft.com/commandline/access-linux-filesystems-in-windows-and-wsl-2/](url)

This was the answer identified by a plotman user as well for making wsl2 quick.  https://github.com/ericaltendorf/plotman/issues/108#issuecomment-822151350
